volatility3/framework/plugins/windows/suspicious_threads.py

```diff
@@ -0,0 +1,201 @@
+# This file is Copyright 2024 Volatility Foundation and licensed under the Volatility Software License 1.0
+# which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
+#
+
+import logging
+from typing import List
+from volatility3.framework import renderers, interfaces
+from volatility3.framework.configuration import requirements
+from volatility3.framework.objects import utility
+from volatility3.framework.renderers import format_hints
+from volatility3.plugins.windows import pslist, threads, vadinfo, thrdscan
+
+vollog = logging.getLogger(__name__)
+
+
+class SupsiciousThreads(interfaces.plugins.PluginInterface):
+    """Lists suspicious userland process threads"""
+
+    _required_framework_version = (2, 4, 0)
+    _version = (2, 0, 0)
+
+    @classmethod
+    def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]:
+        # Since we're calling the plugin, make sure we have the plugin's requirements
+        return [
+            requirements.ModuleRequirement(
+                name="kernel",
+                description="Windows kernel",
+                architectures=["Intel32", "Intel64"],
+            ),
+            requirements.ListRequirement(
+                name="pid",
+                description="Filter on specific process IDs",
+                element_type=int,
+                optional=True,
+            ),
+            requirements.PluginRequirement(
+                name="threads", plugin=threads.Threads, version=(1, 0, 0)
+            ),
+            requirements.VersionRequirement(
+                name="vadinfo", component=vadinfo.VadInfo, version=(2, 0, 0)
+            ),
+        ]
+
+    def _get_ranges(self, kernel, all_ranges, proc):
+        """
+        Maintains a hash table so each process' VADs
+        are only enumerated once per plugin run
+        """
+        key = proc.vol.offset
+
+        if key not in all_ranges:
+            all_ranges[key] = []
+
+            for vad in proc.get_vad_root().traverse():
+                fn = vad.get_file_name()
+                if not isinstance(fn, str) or not fn:
+                    fn = None
+
+                protection_string = vad.get_protection(
+                    vadinfo.VadInfo.protect_values(
+                        self.context, kernel.layer_name, kernel.symbol_table_name
+                    ),
+                    vadinfo.winnt_protections,
+                )
+
+                all_ranges[key].append(
+                    (vad.get_start(), vad.get_end(), protection_string, fn)
+                )
+
+        return all_ranges[key]
+
+    def _get_range(self, ranges, address):
```

Comment:
Typing information please, I know it's a little bit extra, but for simple types it's trivial to add and helps IDEs (and reviewers) figure out how things hang together.  5;)

Reply:
added